src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt

```diff
@@ -2643,6 +2646,123 @@ class TypeSpecTest {
     }.hasMessageThat().isEqualTo("ANNOTATION Taco.eat requires modifiers [PUBLIC, ABSTRACT]")
   }
 
+  @Test fun basicDelegateTest() {
+    val type = TypeSpec.classBuilder("Guac")
+        .primaryConstructor(FunSpec.constructorBuilder()
+            .addParameter("somethingElse", String::class)
+            .build())
+        .addSuperinterface(
+            ParameterizedTypeName.get(Consumer::class, String::class),
+            CodeBlock.of("({ println(it) })"))
+        .build()
+
+    val expect = """
+        |package com.squareup.tacos
+        |
+        |import java.util.function.Consumer
+        |import kotlin.String
+        |
+        |class Guac(somethingElse: String) : Consumer<String> by ({ println(it) })
+        |""".trimMargin()
+
+    assertThat(toString(type)).isEqualTo(expect)
+  }
+
+  @Test fun testMultipleDelegates() {
+    val type = TypeSpec.classBuilder("Guac")
+        .primaryConstructor(FunSpec.constructorBuilder()
+            .addParameter("somethingElse", String::class)
+            .build())
+        .addSuperinterface(ClassName.bestGuess("RandomIface"), CodeBlock.of("Factory.byStrategy(FacadeImpl)"))
+        .addSuperinterface(Runnable::class, CodeBlock.of("Runnable({ Logger.debug(\"Hello world\") })"))
+        .build()
+
+    val expect = """
+        |package com.squareup.tacos
+        |
+        |import java.lang.Runnable
+        |import kotlin.String
+        |
+        |class Guac(somethingElse: String) : RandomIface by Factory.byStrategy(FacadeImpl),
+        |    Runnable by Runnable({ Logger.debug("Hello world") })
+        |""".trimMargin()
+
+    assertThat(toString(type)).isEqualTo(expect)
+  }
+
+  @Test fun testNoSuchParameterDelegate() {
+    assertThrows<IllegalArgumentException> {
+      TypeSpec.classBuilder("Taco")
+          .primaryConstructor(FunSpec.constructorBuilder()
+              .addParameter("other", String::class)
+              .build())
+          .addSuperinterface(KFunction::class, "notOther")
+          .build()
+    }.hasMessageThat().isEqualTo("no such constructor parameter 'notOther' to delegate to for type 'Taco'")
+  }
+
+  @Test fun failAddParamDelegateWhenNullCtor() {
```

Comment:
nit: I don’t like custom abbreviations like `Ctor` because it’s difficult to guess what abbreviation will be used, and that makes grepping more difficult